cmd/kg/webhook.go

```diff
@@ -0,0 +1,276 @@
+	response.Response = &v1.AdmissionResponse{
+		UID: admissionReview.Request.UID,
+	}
+
+	rawExtension := admissionReview.Request.Object
+	var peer kilo.Peer
+
+	if err := json.Unmarshal(rawExtension.Raw, &peer); err != nil {
+		errorCounter.Inc()
+		msg := fmt.Sprintf("could not unmarshal extension to peer spec: %v:", err)
+		level.Error(logger).Log("err", msg)
+		http.Error(w, msg, http.StatusBadRequest)
+		return
+	}
+
+	if err := peer.Validate(); err == nil {
+		level.Debug(logger).Log("msg", "got valid peer spec", "spec", peer.Spec, "name", peer.ObjectMeta.Name)
+		validationCounter.With(prometheus.Labels{"operation": string(admissionReview.Request.Operation), "response": "allowed"}).Inc()
+		response.Response.Allowed = true
+	} else {
+		level.Debug(logger).Log("msg", "got invalid peer spec", "spec", peer.Spec, "name", peer.ObjectMeta.Name)
+		validationCounter.With(prometheus.Labels{"operation": string(admissionReview.Request.Operation), "response": "denied"}).Inc()
+		response.Response.Result = &metav1.Status{
+			Message: err.Error(),
+		}
+	}
+
+	res, err := json.Marshal(response)
+	if err != nil {
+		errorCounter.Inc()
+		msg := fmt.Sprintf("failed to marshal response: %v", err)
+		level.Error(logger).Log("err", msg)
+		http.Error(w, msg, http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	if _, err := w.Write(res); err != nil {
+		level.Error(logger).Log("err", err, "msg", "failed to write response")
+	}
+}
+
+func metricsMiddleWare(path string, next func(http.ResponseWriter, *http.Request)) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		requestCounter.With(prometheus.Labels{"method": r.Method, "handler": path}).Inc()
+		next(w, r)
+	}
+}
+
+func webhook(_ *cobra.Command, _ []string) error {
+	if printVersion {
+		fmt.Println(version.Version)
+		os.Exit(0)
+	}
+	registry.MustRegister(
+		errorCounter,
+		validationCounter,
+		requestCounter,
+	)
+	mm := http.NewServeMux()
+	mm.Handle("/metrics", promhttp.HandlerFor(registry, promhttp.HandlerOpts{}))
+	msrv := &http.Server{
+		Addr:    metricsAddr,
+		Handler: mm,
+	}
+
+	term := make(chan os.Signal, 1)
+	signal.Notify(term, syscall.SIGINT, syscall.SIGTERM)
+
+	exitMetrics := make(chan error, 1)
+	go func() {
+		level.Info(logger).Log("msg", "starting metrics server", "address", msrv.Addr)
+		exitMetrics <- msrv.ListenAndServe()
+	}()
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/validate", metricsMiddleWare("/validate", validationHandler))
+	srv := &http.Server{
+		Addr:    listenAddr,
+		Handler: mux,
+	}
+	exitWebhook := make(chan error, 1)
+	go func() {
+		level.Info(logger).Log("msg", "starting web hook server", "address", srv.Addr)
+		exitWebhook <- srv.ListenAndServeTLS(certPath, keyPath)
+	}()
+
+	ctx, cancel := context.WithCancel(context.Background()) //WithTimeout(context.Background(), 5*time.Second)
+	defer func() {
+		cancel()
+	}()
+
+	select {
```

Comment:
All of this looks totally correct and does pretty much the same things that rungroups we use in kg do. However the rungroups provide some syntactic sugar so they're a little bit easier to maintain, eg you don't have to manually add new cases and bump the count in the wait groups every time that we add a new concurrent thread. I think it might be worth to switch to rungroups for consistency eventually. We can do this later tho